README.md
# go-chess-cli

[![GoDoc](https://godoc.org/github.com/thewizardplusplus/go-chess-cli?status.svg)](https://godoc.org/github.com/thewizardplusplus/go-chess-cli)
[![Go Report Card](https://goreportcard.com/badge/github.com/thewizardplusplus/go-chess-cli)](https://goreportcard.com/report/github.com/thewizardplusplus/go-chess-cli)
[![Build Status](https://travis-ci.org/thewizardplusplus/go-chess-cli.svg?branch=master)](https://travis-ci.org/thewizardplusplus/go-chess-cli)
[![codecov](https://codecov.io/gh/thewizardplusplus/go-chess-cli/branch/master/graph/badge.svg)](https://codecov.io/gh/thewizardplusplus/go-chess-cli)

The chess program with a terminal-based interface.

_**Disclaimer:** this program was written directly on an Android smartphone with the AnGoIde IDE._

## Features

- displaying a board:
  - by symbols (to choose):
    - ASCII;
    - Unicode;
  - by size (to choose):
    - terse;
    - wide;
  - by colors (to choose):
    - monochrome;
    - colorful;
  - misc.:
    - marking searching process;
    - placing a human side at bottom;
- interacting via text commands (moves in [pure algebraic coordinate notation](https://www.chessprogramming.org/Algebraic_Chess_Notation#Pure_coordinate_notation));
- options:
  - initial position in [Forsyth–Edwards notation](https://en.wikipedia.org/wiki/Forsyth–Edwards_Notation);
  - human color (i.e. a computer can move first):
    - support automatic random selecting (optional);
  - move searching restrictions:
    - maximal size of the [transposition table](https://www.chessprogramming.org/Transposition_Table);
    - deep of move searching;
    - duration of move searching;
  - displaying:
    - switching between ASCII/Unicode modes;
    - switching between terse/wide modes;
    - switching between monochrome/colorful modes.

## Installation

```
$ go get github.com/thewizardplusplus/go-chess-cli/...
```

## Usage

```
$ go-chess-cli -h | -help | --help
$ go-chess-cli [options]
```

Options:

- `-h`, `-help`, `--help` &mdash; show the help message and exit;
- `-cacheSize ITEMS` &mdash; maximal cache size (default: `1000000`, i.e. one million);
- `-colorfulBoard {false|true}` &mdash; use colors to display the board (default: `true`; for inverting use `-colorfulBoard=false`);
- `-colorfulPieces {false|true}` &mdash; use colors to display pieces (default: `true`; for inverting use `-colorfulPieces=false`);
- `-deep INTEGER` &mdash; search deep (default: `5`);
- `-duration DURATION` &mdash; search duration (e.g. `72h3m0.5s`; default: `5s`);
- `-fen STRING` &mdash; board in FEN (default: `rnbqk/ppppp/5/PPPPP/RNBQK`, i.e. Gardner's minichess);
- `-humanColor {random|black|white}` &mdash; human color (default: `random`);
- `-pieceBlackColor INTEGER` &mdash; SGR parameter for ANSI escape sequences for setting a color of black pieces (default: `34`; see for details: https://en.wikipedia.org/wiki/ANSI_escape_code#3/4_bit);
- `-pieceWhiteColor INTEGER` &mdash; SGR parameter for ANSI escape sequences for setting a color of white pieces (default: `31`; see for details: https://en.wikipedia.org/wiki/ANSI_escape_code#3/4_bit);
- `-squareBlackColor INTEGER` &mdash; SGR parameter for ANSI escape sequences for setting a color of black squares (default: `40`; see for details: https://en.wikipedia.org/wiki/ANSI_escape_code#3/4_bit);
- `-squareWhiteColor INTEGER` &mdash; SGR parameter for ANSI escape sequences for setting a color of white squares (default: `47`; see for details: https://en.wikipedia.org/wiki/ANSI_escape_code#3/4_bit);
- `-unicode {false|true}` &mdash; use Unicode to display pieces (default: `true`; for inverting use `-unicode=false`);
- `-wide {false|true}` &mdash; display the board wide (default: `true`; for inverting use `-wide=false`).

## Examples

`ascii.DecodeColor()`:

```go
package main

import (
	"fmt"

	"github.com/thewizardplusplus/go-chess-cli/encoding/ascii"
)

func main() {
	color, _ := ascii.DecodeColor("white")
	fmt.Printf("%v\n", color)

	// Output: 1
}
```

`ascii.EncodeColor()`:

```go
package main

import (
	"fmt"

	"github.com/thewizardplusplus/go-chess-cli/encoding/ascii"
	models "github.com/thewizardplusplus/go-chess-models"
)

func main() {
	color := ascii.EncodeColor(models.White)
	fmt.Printf("%v\n", color)

	// Output: white
}
```

`ascii.PieceStorageEncoder.EncodePieceStorage()`:

```go
package main

import (
	"fmt"

	"github.com/thewizardplusplus/go-chess-cli/encoding/ascii"
	models "github.com/thewizardplusplus/go-chess-models"
	"github.com/thewizardplusplus/go-chess-models/encoding/uci"
	"github.com/thewizardplusplus/go-chess-models/pieces"
)

func main() {
	const fen = "r3k2r/p1ppqpb1/bn2pnp1/3PN3/1p2P3/2N2Q1p/PPPBBPPP/R3K2R"
	storage, _ := uci.DecodePieceStorage(fen, pieces.NewPiece, models.NewBoard)
	encoder := ascii.NewPieceStorageEncoder(
		uci.EncodePiece,
		"x",
		ascii.Margins{},
		ascii.WithoutColor,
		models.Black,
		1,
	)
	fmt.Printf("%v\n", encoder.EncodePieceStorage(storage))

	// Output:
	// 8rxxxkxxr
	// 7pxppqpbx
	// 6bnxxpnpx
	// 5xxxPNxxx
	// 4xpxxPxxx
	// 3xxNxxQxp
	// 2PPPBBPPP
	// 1RxxxKxxR
	//  abcdefgh
}
```

`ascii.PieceStorageEncoder.EncodePieceStorage()` with margins:

```go
package main

import (
	"fmt"

	"github.com/thewizardplusplus/go-chess-cli/encoding/ascii"
	models "github.com/thewizardplusplus/go-chess-models"
	"github.com/thewizardplusplus/go-chess-models/encoding/uci"
	"github.com/thewizardplusplus/go-chess-models/pieces"
)

func main() {
	margins := ascii.Margins{
		Piece: ascii.PieceMargins{
			HorizontalMargins: ascii.HorizontalMargins{
				Left: 1,
			},
		},
		Legend: ascii.LegendMargins{
			Rank: ascii.HorizontalMargins{
				Right: 1,
			},
		},
	}

	const fen = "r3k2r/p1ppqpb1/bn2pnp1/3PN3/1p2P3/2N2Q1p/PPPBBPPP/R3K2R"
	storage, _ := uci.DecodePieceStorage(fen, pieces.NewPiece, models.NewBoard)
	encoder := ascii.NewPieceStorageEncoder(
		uci.EncodePiece,
		"x",
		margins,
		ascii.WithoutColor,
		models.Black,
		1,
	)
	fmt.Printf("%v\n", encoder.EncodePieceStorage(storage))

	// Output:
	// 8  r x x x k x x r
	// 7  p x p p q p b x
	// 6  b n x x p n p x
	// 5  x x x P N x x x
	// 4  x p x x P x x x
	// 3  x x N x x Q x p
	// 2  P P P B B P P P
	// 1  R x x x K x x R
	//    a b c d e f g h
}
```

`ascii.PieceStorageEncoder.EncodePieceStorage()` with colors:

```go
package main

import (
	"fmt"

	"github.com/thewizardplusplus/go-chess-cli/encoding/ascii"
	climodels "github.com/thewizardplusplus/go-chess-cli/models"
	models "github.com/thewizardplusplus/go-chess-models"
	"github.com/thewizardplusplus/go-chess-models/encoding/uci"
	"github.com/thewizardplusplus/go-chess-models/pieces"
)

func main() {
	colorizer := func(text string, color climodels.OptionalColor) string {
		var colorMark byte
		if color.IsSet {
			colorMark = ascii.EncodeColor(color.Value)[0]
		} else {
			colorMark = 'n'
		}

		return fmt.Sprintf("(%c%s)", colorMark, text)
	}

	const fen = "r3k2r/p1ppqpb1/bn2pnp1/3PN3/1p2P3/2N2Q1p/PPPBBPPP/R3K2R"
	storage, _ := uci.DecodePieceStorage(fen, pieces.NewPiece, models.NewBoard)
	encoder := ascii.NewPieceStorageEncoder(
		uci.EncodePiece,
		"x",
		ascii.Margins{},
		colorizer,
		models.Black,
		1,
	)
	fmt.Printf("%v\n", encoder.EncodePieceStorage(storage))

	// Output:
	// (n8)(wr)(bx)(wx)(bx)(wk)(bx)(wx)(br)
	// (n7)(bp)(wx)(bp)(wp)(bq)(wp)(bb)(wx)
	// (n6)(wb)(bn)(wx)(bx)(wp)(bn)(wp)(bx)
	// (n5)(bx)(wx)(bx)(wP)(bN)(wx)(bx)(wx)
	// (n4)(wx)(bp)(wx)(bx)(wP)(bx)(wx)(bx)
	// (n3)(bx)(wx)(bN)(wx)(bx)(wQ)(bx)(wp)
	// (n2)(wP)(bP)(wP)(bB)(wB)(bP)(wP)(bP)
	// (n1)(bR)(wx)(bx)(wx)(bK)(wx)(bx)(wR)
	// (n )(na)(nb)(nc)(nd)(ne)(nf)(ng)(nh)
}
```

`unicode.EncodePiece()`:

```go
package main

import (
	"fmt"

	"github.com/thewizardplusplus/go-chess-cli/encoding/unicode"
	models "github.com/thewizardplusplus/go-chess-models"
	"github.com/thewizardplusplus/go-chess-models/pieces"
)

func main() {
	fen := unicode.EncodePiece(pieces.NewBishop(models.White, models.Position{}))
	fmt.Printf("%v\n", fen)

	// Output: ♗
}
```

## License

The MIT License (MIT)

Copyright &copy; 2019-2020 thewizardplusplus


CHANGE_LOG.md
# Change Log

## [v1.3](https://github.com/thewizardplusplus/go-chess-cli/tree/v1.3) (2020-02-21)

- fixing the code style;
- adding usage examples;
- improving:
  - repository decor;
  - CI configuration.

## [v1.2](https://github.com/thewizardplusplus/go-chess-cli/tree/v1.2) (2020-02-09)

- displaying:
  - in the colorful mode:
    - pieces (optional);
    - board (optional);
  - fix displaying a board in the wide mode.

## [v1.1](https://github.com/thewizardplusplus/go-chess-cli/tree/v1.1) (2020-01-15)

- displaying:
  - pieces by Unicode (optional);
  - board in the wide mode (optional);
  - misc.:
    - marking searching process;
    - placing a human side at bottom;
- automatic random selecting a human color (optional);
- restricting move searching by a deep.

## [v1.0](https://github.com/thewizardplusplus/go-chess-cli/tree/v1.0) (2019-11-02)


cmd/go-chess-cli/main.go
package main

import (
	"bufio"
	"errors"
	"flag"
	"fmt"
	"io"
	"log"
	"math/rand"
	"os"
	"runtime"
	"strings"
	"time"

	"github.com/thewizardplusplus/go-chess-cli/encoding/ascii"
	"github.com/thewizardplusplus/go-chess-cli/encoding/unicode"
	climodels "github.com/thewizardplusplus/go-chess-cli/models"
	minimax "github.com/thewizardplusplus/go-chess-minimax"
	"github.com/thewizardplusplus/go-chess-minimax/caches"
	"github.com/thewizardplusplus/go-chess-minimax/evaluators"
	moves "github.com/thewizardplusplus/go-chess-minimax/models"
	"github.com/thewizardplusplus/go-chess-minimax/terminators"
	models "github.com/thewizardplusplus/go-chess-models"
	"github.com/thewizardplusplus/go-chess-models/encoding/uci"
	"github.com/thewizardplusplus/go-chess-models/pieces"
)

// nolint: gochecknoglobals
var (
	baseWideMargins = ascii.Margins{
		Legend: ascii.LegendMargins{
			File: ascii.VerticalMargins{
				Top: 1,
			},
			Rank: ascii.HorizontalMargins{
				Right: 1,
			},
		},
		Board: ascii.VerticalMargins{
			Top:    1,
			Bottom: 1,
		},
	}
	widePieceMargins = ascii.PieceMargins{
		HorizontalMargins: ascii.HorizontalMargins{
			Left: 1,
		},
		VerticalMargins: ascii.VerticalMargins{
			Bottom: 1,
		},
	}
	extraWidePieceMargins = ascii.PieceMargins{
		HorizontalMargins: ascii.HorizontalMargins{
			Left:  1,
			Right: 1,
		},
		VerticalMargins: ascii.VerticalMargins{
			Top:    1,
			Bottom: 1,
		},
	}
)

type colorCodeGroup map[models.Color]int

func setTTYMode(mode int) string {
	return fmt.Sprintf("\x1b[%dm", mode)
}

func makeColorizer(colorsCodes colorCodeGroup) ascii.Colorizer {
	return func(text string, color models.Color) string {
		return setTTYMode(colorsCodes[color]) + text + setTTYMode(0)
	}
}

func search(
	cache caches.Cache,
	storage models.PieceStorage,
	color models.Color,
	terminator terminators.SearchTerminator,
) (moves.ScoredMove, error) {
	searcher := minimax.NewParallelSearcher(
		terminator,
		runtime.NumCPU(),
		func() minimax.MoveSearcher {
			innerSearcher := minimax.NewAlphaBetaSearcher(
				models.MoveGenerator{},
				nil, // terminator will be set automatically by the iterative searcher
				evaluators.MaterialEvaluator{},
			)

			if cache != nil {
				// make and bind a cached searcher to inner one
				minimax.NewCachedSearcher(innerSearcher, cache)
			}

			return minimax.NewIterativeSearcher(
				innerSearcher,
				nil, // terminator will be set automatically by the parallel searcher
			)
		},
	)

	return searcher.SearchMove(
		storage,
		color,
		0, // initial deep
		moves.NewBounds(),
	)
}

func check(storage models.PieceStorage, color models.Color) error {
	// minimal deep, at which a game state will be detected
	terminator := terminators.NewDeepTerminator(1)
	_, err := search(
		nil, // without a cache
		storage,
		color,
		terminator,
	)
	return err // don't wrap
}

func writePrompt(
	storageEncoder ascii.PieceStorageEncoder,
	storage models.PieceStorage,
	color models.Color,
	side climodels.Side,
) error {
	text := storageEncoder.EncodePieceStorage(storage)
	fmt.Println(text)

	if err := check(storage, color); err != nil {
		return err // don't wrap
	}

	var mark string
	if side == climodels.Searcher {
		mark = "(searching) "
	}

	text = ascii.EncodeColor(color)
	fmt.Printf("%s> %s", text, mark) // don't break the line

	return nil
}

func readMove(
	reader *bufio.Reader,
	storageEncoder ascii.PieceStorageEncoder,
	storage models.PieceStorage,
	color models.Color,
	side climodels.Side,
) (models.Move, error) {
	if err := writePrompt(storageEncoder, storage, color, side); err != nil {
		return models.Move{}, err // don't wrap
	}

	text, err := reader.ReadString('\n')
	if err != nil && err != io.EOF {
		return models.Move{}, fmt.Errorf("unable to read the move: %s", err)
	}

	text = strings.TrimSuffix(text, "\n")
	move, err := uci.DecodeMove(text)
	if err != nil {
		return models.Move{}, fmt.Errorf("unable to decode the move: %s", err)
	}

	if err = storage.CheckMove(move); err != nil {
		return models.Move{}, fmt.Errorf("incorrect move: %s", err)
	}

	if piece, _ := storage.Piece(move.Start); piece.Color() != color {
		return models.Move{}, errors.New("incorrect move: opponent piece")
	}

	nextStorage := storage.ApplyMove(move)
	nextColor := color.Negative()
	if err = check(nextStorage, nextColor); err == models.ErrKingCapture {
		return models.Move{}, errors.New("incorrect move: check")
	}

	return move, nil
}

func searchMove(
	cache caches.Cache,
	storageEncoder ascii.PieceStorageEncoder,
	storage models.PieceStorage,
	color models.Color,
	side climodels.Side,
	deep int,
	duration time.Duration,
) (models.Move, error) {
	if err := writePrompt(storageEncoder, storage, color, side); err != nil {
		return models.Move{}, err // don't wrap
	}

	terminator := terminators.NewGroupTerminator(
		terminators.NewDeepTerminator(deep),
		terminators.NewTimeTerminator(time.Now, duration),
	)
	move, _ := search(cache, storage, color, terminator) // nolint: gosec
	return move.Move, nil
}

func main() {
	rand.Seed(time.Now().UnixNano())

	fen := flag.String(
		"fen",
		"rnbqk/ppppp/5/PPPPP/RNBQK",
		"board in FEN (default: Gardner's minichess)",
	)
	humanColor := flag.String(
		"humanColor",
		"random",
		"human color (allowed: random, black, white)",
	)
	deep := flag.Int("deep", 5, "search deep")
	duration := flag.Duration(
		"duration",
		5*time.Second,
		"search duration (e.g. 72h3m0.5s)",
	)
	cacheSize := flag.Int("cacheSize", 1e6, "maximal cache size (in items)")
	useUnicode := flag.Bool("unicode", true, "use Unicode to display pieces")
	colorfulPieces := flag.Bool(
		"colorfulPieces",
		true,
		"use colors to display pieces",
	)
	pieceBlackColor := flag.Int(
		"pieceBlackColor",
		34, // blue
		"SGR parameter for ANSI escape sequences for setting a color of black pieces",
	)
	pieceWhiteColor := flag.Int(
		"pieceWhiteColor",
		31, // red
		"SGR parameter for ANSI escape sequences for setting a color of white pieces",
	)
	colorfulBoard := flag.Bool(
		"colorfulBoard",
		true,
		"use colors to display the board",
	)
	squareBlackColor := flag.Int(
		"squareBlackColor",
		40, // black
		"SGR parameter for ANSI escape sequences "+
			"for setting a color of black squares",
	)
	squareWhiteColor := flag.Int(
		"squareWhiteColor",
		47, // white
		"SGR parameter for ANSI escape sequences "+
			"for setting a color of white squares",
	)
	wide := flag.Bool("wide", true, "display the board wide")
	flag.Parse()

	storage, err := uci.DecodePieceStorage(*fen, pieces.NewPiece, models.NewBoard)
	if err != nil {
		log.Fatal("unable to decode the board: ", err)
	}

	parsedHumanColor, err := ascii.DecodeColor(*humanColor)
	switch {
	case err == nil:
	case *humanColor == "random":
		if rand.Intn(2) == 0 {
			parsedHumanColor = models.Black
		} else {
			parsedHumanColor = models.White
		}
	default:
		log.Fatal("unable to decode the color: ", err)
	}

	var pieceEncoder ascii.PieceEncoder
	var placeholder string
	if *useUnicode {
		pieceEncoder = unicode.EncodePiece
		placeholder = "\u00b7"
	} else {
		pieceEncoder = uci.EncodePiece
		placeholder = "."
	}
	if *colorfulPieces {
		pieceColorizer := makeColorizer(colorCodeGroup{
			models.Black: *pieceBlackColor,
			models.White: *pieceWhiteColor,
		})
		basePieceEncoder := pieceEncoder
		pieceEncoder = func(piece models.Piece) string {
			text := basePieceEncoder(piece)
			return pieceColorizer(text, piece.Color())
		}
	}
	if *colorfulBoard {
		placeholder = " "
	}

	var margins ascii.Margins
	if *wide {
		margins = baseWideMargins

		if *colorfulBoard {
			margins.Piece = extraWidePieceMargins
		} else {
			margins.Piece = widePieceMargins
		}
	}

	var squareColorizer ascii.OptionalColorizer
	if *colorfulBoard {
		baseSquareColorizer := makeColorizer(colorCodeGroup{
			models.Black: *squareBlackColor,
			models.White: *squareWhiteColor,
		})
		squareColorizer = ascii.NewOptionalColorizer(baseSquareColorizer)
	} else {
		squareColorizer = ascii.WithoutColor
	}

	side := climodels.NewSide(parsedHumanColor)
	reader := bufio.NewReader(os.Stdin)
	storageEncoder := ascii.NewPieceStorageEncoder(
		pieceEncoder,
		placeholder,
		margins,
		squareColorizer,
		parsedHumanColor.Negative(),
		1,
	)
	cache := caches.NewParallelCache(caches.NewStringHashingCache(
		*cacheSize,
		uci.EncodePieceStorage,
	))
loop:
	for {
		var move models.Move
		var err error
		switch side {
		case climodels.Human:
			move, err = readMove(reader, storageEncoder, storage, parsedHumanColor, side)
		case climodels.Searcher:
			move, err = searchMove(
				cache,
				storageEncoder,
				storage,
				parsedHumanColor.Negative(),
				side,
				*deep,
				*duration,
			)
			if err == nil {
				text := uci.EncodeMove(move)
				fmt.Println(text)
			}
		}
		switch err {
		case nil:
		case minimax.ErrCheckmate, minimax.ErrDraw:
			log.Print("game in the state: ", err)
			break loop
		default:
			log.Print("error: ", err)
			continue loop
		}

		storage = storage.ApplyMove(move)
		side = side.Invert()
	}
}
